main.cpp

```diff
@@ -288,436 +236,179 @@ struct HeapA
-    return *this;
-}
-// ======================================== Float operators ===========
-FloatType::operator float() const { return *value; }
-
-FloatType& FloatType::operator+=(float rhs)
-{
-    *value += rhs;
-    return *this;
-}
+    explicit Numeric(Type type_) : value(std::make_unique<Type>(type_)) { }
+    ~Numeric() { value.reset(nullptr); }
+    //========== operators ==========
+    operator Type() const { return *value; }
 
-FloatType& FloatType::operator-=(float rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator*=(float rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator/=(float rhs)
-{
-    if (rhs == 0.0f)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-    *value /= rhs;
-    return *this;
-}
-// ======================================== Float Apply func ==========
-FloatType& FloatType::apply(std::function<FloatType&(float&)> f)
-{
-    if (f)
+    Numeric<Type>& operator+=(Type rhs)
     {
-        return f(*value);
+        *value += rhs;
+        return *this;
     }
-    
-    return *this;
-}
 
-FloatType& FloatType::apply(void (*f)(float&))
-{
-    if (f)
+    Numeric<Type>& operator-=(Type rhs)
     {
-        f(*value);
+        *value -= rhs;
+        return *this;
     }
-    
-    return *this;
-}
-// ======================================== Double primitive ==========
-DoubleType& DoubleType::add(double rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::subtract(double rhs)
-{
-    *value -= rhs;
-    return *this;
-}
 
-DoubleType& DoubleType::multiply(double rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::divide(double rhs)
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-
-    *value /= rhs;
-    return *this;        
-}
-// ======================================== Double operators ==========
-DoubleType::operator double() const { return *value; }
-
-DoubleType& DoubleType::operator+=(double rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator-=(double rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator*=(double rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator/=(double rhs)
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-
-    *value /= rhs;
-    return *this;        
-}
-// ======================================== Double Apply func =========
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> f)
-{
-    if (f)
+    Numeric<Type>& operator*=(Type rhs)
     {
-        return f(*value);
+        *value *= rhs;
+        return *this;
     }
-    
-    return *this;
-}
 
-DoubleType& DoubleType::apply(void (*f)(double&))
-{
-    if (f)
+    template <typename ArgumentType>
+    Numeric<Type>& operator/=(ArgumentType rhs)
```

Comment:
Inside the templated class, you do not need to write `Numeric<Type>&` as the return type.
`Numeric&` is all you need to write.
Your function argument should be passed by const reference to avoid unnecessary copies.
- [ ] revise